{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6008c76d",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "params_resource_dir = \"../resources/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3cd6dab",
   "metadata": {},
   "source": [
    "# Datasets - General datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d3df2b9",
   "metadata": {},
   "source": [
    "## Overview"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17d9cdd2",
   "metadata": {},
   "source": [
    "Manipulating Genome annotations and FASTA files.\n",
    "\n",
    "\n",
    "- [Genome data](#genome-data)\n",
    "- [Full example: from a transcript of interest to extracted CDS sequences](#integrating-everything-intervals-and-sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f19cbae",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "498d9fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dataclasses\n",
    "import os\n",
    "import sys\n",
    "from io import StringIO\n",
    "from numbers import Number\n",
    "from pathlib import Path\n",
    "from typing import Literal\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pybedtools as pbt\n",
    "import pyBigWig as pbw\n",
    "import pydantic.dataclasses\n",
    "import pyfaidx\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from pydantic.dataclasses import dataclass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a00b08a",
   "metadata": {},
   "source": [
    "## Initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0358c5e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using resources from: ../resources\n"
     ]
    }
   ],
   "source": [
    "resource_dir = Path(params_resource_dir)\n",
    "if not resource_dir.exists():\n",
    "    raise FileNotFoundError(\"Resource directory does not exist: \" + str(resource_dir))\n",
    "\n",
    "print(\"Using resources from:\", resource_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47228cd9",
   "metadata": {},
   "source": [
    "## Genome data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a984d3e1",
   "metadata": {},
   "source": [
    "### Genome chromosome sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ccdeb224",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading the \"genome file\" from: ../resources/general/genome/hg38.chrom.sizes\n",
      "Main chromosomes: chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrX, chrY, chrM\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248_956_422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>242_193_529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3</td>\n",
       "      <td>198_295_559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>190_214_555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>181_538_259</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom         size\n",
       "0  chr1  248_956_422\n",
       "1  chr2  242_193_529\n",
       "2  chr3  198_295_559\n",
       "3  chr4  190_214_555\n",
       "4  chr5  181_538_259"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of positions from the main chromosomes: 3_088_286_401\n",
      "\n",
      "\n",
      "Number of additional chromosome regions: 430\n",
      "Examples: chr10_GL383545v1_alt, chr10_GL383546v1_alt, chr10_KI270824v1_alt\n"
     ]
    }
   ],
   "source": [
    "filepath = resource_dir / \"general\" / \"genome\" / \"hg38.chrom.sizes\"\n",
    "print('Loading the \"genome file\" from:', filepath)\n",
    "\n",
    "genome_file = pd.read_csv(filepath, header=None, sep=\"\\t\")\n",
    "genome_file.columns = [\"chrom\", \"size\"]\n",
    "\n",
    "main_autosomes = [f\"chr{i}\" for i in range(1, 23)]\n",
    "sex_chromosomes = [\"chrX\", \"chrY\"]\n",
    "mitochondrial_chromosome = [\"chrM\"]\n",
    "main_chromosomes = list(set(main_autosomes) | set(sex_chromosomes) | set(mitochondrial_chromosome))\n",
    "\n",
    "print(f\"Main chromosomes: {', '.join(main_autosomes + sex_chromosomes + mitochondrial_chromosome)}\")\n",
    "print(\"\\n\")\n",
    "\n",
    "display(genome_file.head(5).assign(size=lambda x: x[\"size\"].map(\"{:_}\".format)))\n",
    "\n",
    "print(\n",
    "    f\"Total number of positions from the main chromosomes: {genome_file.loc[genome_file['chrom'].isin(main_chromosomes), 'size'].sum():_}\"\n",
    ")\n",
    "\n",
    "print(\"\\n\")\n",
    "\n",
    "other_chromosomes = sorted(list(set(genome_file.chrom) - set(main_chromosomes)))\n",
    "\n",
    "print(f\"Number of additional chromosome regions: {len(other_chromosomes):_}\")\n",
    "print(\"Examples: \" + \", \".join(other_chromosomes[:3]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "993e03a1",
   "metadata": {},
   "source": [
    "### Genome annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e0d8b17",
   "metadata": {},
   "source": [
    "The genome has been annotated with multiple features and properties, which can be \n",
    "represented as genomic intervals, i.e. the coordinate system with :\n",
    "\n",
    "- chromosome\n",
    "- start position\n",
    "- end position\n",
    "\n",
    "Additionally, the **strand** ('plus' or 'minus', represented as '+' and '-' or as integers `1` and `-1`) can be specified.\n",
    "\n",
    "Importantly: the genomic intervals are usually represented as 0-based, half-open intervals,\n",
    "but this depends on the file format and the convention adopted by the \n",
    "database."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "271d2b86",
   "metadata": {},
   "source": [
    "Gene annotations are the most obvious example of annotations: defining the \n",
    "genomic intervals associated with an experimentally characterized entity.\n",
    "\n",
    "\n",
    "Multiple annotations can overlap:\n",
    "\n",
    "- either on the same strand: \n",
    "    - this may be the case for sub-annotations related to a \"meta\" annotation ;\n",
    "    - e.g. a gene can actually be *transcribed* into multiple *transcripts*.\n",
    "    - e.g. a transcript can be \"segmented\" into *exons* (implicitly separated by *introns*), further segmented into *CDS* intervals and *UTR* (UnTranslated Region) intervals.\n",
    "\n",
    "- or on different strands: biologically, the two strands can bear genes. Furthermore, specific annotations (e.g. bias in sequence content) can be strand-specific.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01bbce31",
   "metadata": {},
   "source": [
    "Multiple online reference resources can provide with different annotations for the same genome.\n",
    "\n",
    "E.g. different resources apply different criteria to define the same gene.\n",
    "\n",
    "The provided resources for gene annotations are here:\n",
    "\n",
    "- GENCODE: a comprehensive annotation of the (human) genome.\n",
    "- MANE: a high-quality curated set of human transcripts, that attempts to provide a single representative transcript for each gene, that is consistent with different resources (including GENCODE)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5018eb0",
   "metadata": {},
   "source": [
    "Different data formats exist to represent genomic annotations within a genome:\n",
    "\n",
    "- **General Feature Format** and **GTF** (General Transfer Format): tabular format, where genomic intervals are 1-based, closed intervals (i.e. the first position is 1, and the end position is included in the interval).\n",
    "    - Different versions of the GFF format exist, the latest being GFF3.\n",
    "    - GTF is a specific version of GFF.\n",
    "\n",
    "- **BED** (Browser Extensible Data): tabular format, where genomic intervals are 0-based, half-open intervals (i.e. the first position is 0, and the end position is not included in the interval).\n",
    "    - Commonly represented as a 3-column (chrom, start, end), 4 column (chrom, start, end, name) or 6-column (chrom, start, end, name, score, strand) format.\n",
    "    - Additional columns can be added (up to 12) to represent additional information, notably for visualization purposes.\n",
    "\n",
    "Such file formats can be compressed with `bgzip` (a block compression format) and indexed with `tabix` (a tool to index tabular data), and can be queried with specific libraries to enable fast access to specific genomic intervals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8a72f11b",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed6_cols = [\"chrom\", \"start\", \"end\", \"name\", \"score\", \"strand\"]\n",
    "gff_columns = [\n",
    "    \"seqid\",\n",
    "    \"source\",\n",
    "    \"biotype\",\n",
    "    \"start\",\n",
    "    \"end\",\n",
    "    \"score\",\n",
    "    \"strand\",\n",
    "    \"phase\",\n",
    "    \"attributes\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba10a8f3",
   "metadata": {},
   "source": [
    "#### Example - MANE gene annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2f69b49f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>biotype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000142611.17;gene_id=ENSG00000142611....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3069203</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000270722.10;Parent=ENSG00000142611.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3069203</td>\n",
       "      <td>3069296</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000270722.10:1;Parent=ENST000002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3069260</td>\n",
       "      <td>3069296</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000270722.10;Parent=ENST000002707...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>3069260</td>\n",
       "      <td>3069262</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=start_codon:ENST00000270722.10;Parent=ENST0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid          source      biotype    start      end score strand phase  \\\n",
       "0  chr1  ensembl_havana         gene  3069168  3438621     .      +     .   \n",
       "1  chr1  ensembl_havana   transcript  3069203  3438621     .      +     .   \n",
       "2  chr1  ensembl_havana         exon  3069203  3069296     .      +     .   \n",
       "3  chr1  ensembl_havana          CDS  3069260  3069296     .      +     0   \n",
       "4  chr1  ensembl_havana  start_codon  3069260  3069262     .      +     0   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000142611.17;gene_id=ENSG00000142611....  \n",
       "1  ID=ENST00000270722.10;Parent=ENSG00000142611.1...  \n",
       "2  ID=exon:ENST00000270722.10:1;Parent=ENST000002...  \n",
       "3  ID=CDS:ENST00000270722.10;Parent=ENST000002707...  \n",
       "4  ID=start_codon:ENST00000270722.10;Parent=ENST0...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filepath = resource_dir / \"general\" / \"gene_annotations.hg38.MANE_v1.4\" / \"MANE.GRCh38.v1.4.ensembl_genomic.gff.gz\"\n",
    "\n",
    "tmp_gff = pd.read_csv(filepath, sep=\"\\t\", compression=\"gzip\", header=None, comment=\"#\", nrows=10, names=gff_columns)\n",
    "display(tmp_gff.head(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91f3dbd7",
   "metadata": {},
   "source": [
    "Two BED files are provided, extracted from the GFF file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f1c7e4aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65418</td>\n",
       "      <td>71585</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450739</td>\n",
       "      <td>451678</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685715</td>\n",
       "      <td>686654</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923922</td>\n",
       "      <td>944575</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944202</td>\n",
       "      <td>959309</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom   start     end                name score strand\n",
       "0  chr1   65418   71585   ENSG00000186092.7     .      +\n",
       "1  chr1  450739  451678   ENSG00000284733.2     .      -\n",
       "2  chr1  685715  686654   ENSG00000284662.2     .      -\n",
       "3  chr1  923922  944575  ENSG00000187634.13     .      +\n",
       "4  chr1  944202  959309  ENSG00000188976.11     .      -"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65418</td>\n",
       "      <td>71585</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450739</td>\n",
       "      <td>451678</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685715</td>\n",
       "      <td>686654</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923922</td>\n",
       "      <td>944574</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944202</td>\n",
       "      <td>959256</td>\n",
       "      <td>ENST00000327044.7</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom   start     end               name score strand\n",
       "0  chr1   65418   71585  ENST00000641515.2     .      +\n",
       "1  chr1  450739  451678  ENST00000426406.4     .      -\n",
       "2  chr1  685715  686654  ENST00000332831.5     .      -\n",
       "3  chr1  923922  944574  ENST00000616016.5     .      +\n",
       "4  chr1  944202  959256  ENST00000327044.7     .      -"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filepath = (\n",
    "    resource_dir / \"general\" / \"gene_annotations.hg38.MANE_v1.4\" / \"MANE.GRCh38.v1.4.ensembl_genomic.genes.bed.gz\"\n",
    ")\n",
    "tmp1 = pd.read_csv(\n",
    "    filepath,\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\",\n",
    "    header=None,\n",
    "    comment=\"#\",\n",
    "    nrows=100,\n",
    "    names=bed6_cols,\n",
    ")\n",
    "display(tmp1.head(5))\n",
    "\n",
    "filepath = (\n",
    "    resource_dir / \"general\" / \"gene_annotations.hg38.MANE_v1.4\" / \"MANE.GRCh38.v1.4.ensembl_genomic.transcripts.bed.gz\"\n",
    ")\n",
    "tmp2 = pd.read_csv(\n",
    "    filepath,\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\",\n",
    "    header=None,\n",
    "    comment=\"#\",\n",
    "    nrows=100,\n",
    "    names=bed6_cols,\n",
    ")\n",
    "display(tmp2.head(5))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2d4fe20",
   "metadata": {},
   "source": [
    "Intervals can be intersected with the `BedTool` library (and its `pyBedTools` Python wrapper)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e00a9df6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a.chrom</th>\n",
       "      <th>a.start</th>\n",
       "      <th>a.end</th>\n",
       "      <th>a.name</th>\n",
       "      <th>a.score</th>\n",
       "      <th>a.strand</th>\n",
       "      <th>b.chrom</th>\n",
       "      <th>b.start</th>\n",
       "      <th>b.end</th>\n",
       "      <th>b.name</th>\n",
       "      <th>b.score</th>\n",
       "      <th>b.strand</th>\n",
       "      <th>overlap_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65418</td>\n",
       "      <td>71585</td>\n",
       "      <td>ENSG00000186092.7</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>65418</td>\n",
       "      <td>71585</td>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>6167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450739</td>\n",
       "      <td>451678</td>\n",
       "      <td>ENSG00000284733.2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>450739</td>\n",
       "      <td>451678</td>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685715</td>\n",
       "      <td>686654</td>\n",
       "      <td>ENSG00000284662.2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>685715</td>\n",
       "      <td>686654</td>\n",
       "      <td>ENST00000332831.5</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923922</td>\n",
       "      <td>944575</td>\n",
       "      <td>ENSG00000187634.13</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>923922</td>\n",
       "      <td>944574</td>\n",
       "      <td>ENST00000616016.5</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>20652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944202</td>\n",
       "      <td>959309</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>944202</td>\n",
       "      <td>959256</td>\n",
       "      <td>ENST00000327044.7</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>15054</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  a.chrom  a.start   a.end              a.name a.score a.strand b.chrom  \\\n",
       "0    chr1    65418   71585   ENSG00000186092.7       .        +    chr1   \n",
       "1    chr1   450739  451678   ENSG00000284733.2       .        -    chr1   \n",
       "2    chr1   685715  686654   ENSG00000284662.2       .        -    chr1   \n",
       "3    chr1   923922  944575  ENSG00000187634.13       .        +    chr1   \n",
       "4    chr1   944202  959309  ENSG00000188976.11       .        -    chr1   \n",
       "\n",
       "   b.start   b.end             b.name b.score b.strand  overlap_size  \n",
       "0    65418   71585  ENST00000641515.2       .        +          6167  \n",
       "1   450739  451678  ENST00000426406.4       .        -           939  \n",
       "2   685715  686654  ENST00000332831.5       .        -           939  \n",
       "3   923922  944574  ENST00000616016.5       .        +         20652  \n",
       "4   944202  959256  ENST00000327044.7       .        -         15054  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "a.name\n",
       "ENSG00000041988.16    1\n",
       "ENSG00000007923.17    1\n",
       "ENSG00000171735.21    1\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "b.name\n",
       "ENST00000054650.9     1\n",
       "ENST00000377577.10    1\n",
       "ENST00000303635.12    1\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bt_tmp1 = pbt.BedTool.from_dataframe(tmp1)\n",
    "bt_tmp2 = pbt.BedTool.from_dataframe(tmp2)\n",
    "\n",
    "intersect = bt_tmp1.intersect(bt_tmp2, s=True, wao=True).to_dataframe(\n",
    "    names=[\"a.\" + c for c in bed6_cols] + [\"b.\" + c for c in bed6_cols] + [\"overlap_size\"]\n",
    ")\n",
    "\n",
    "display(intersect.head(5))\n",
    "\n",
    "if any(intersect[\"a.name\"].value_counts() != 1):\n",
    "    raise ValueError(\"The gene IDs column has non-unique values. This is unexpected.\")\n",
    "\n",
    "display(intersect[\"a.name\"].value_counts().sort_values().head(3))\n",
    "\n",
    "\n",
    "if any(intersect[\"b.name\"].value_counts() != 1):\n",
    "    raise ValueError(\"The transcript IDs column has non-unique values. This is unexpected.\")\n",
    "\n",
    "display(intersect[\"b.name\"].value_counts().sort_values().head(3))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb47d9ac",
   "metadata": {},
   "source": [
    "Here we have intersected the pre-extracted `biotype=\"gene\"` annotations labeled with their gene IDs with the pre-extracted `biotype=\"transcript\"` annotations labeled with their transcript IDs.\n",
    "\n",
    "The `value_counts()` operation highlights that the MANE dataset is curated, providing with a single transcript for each gene."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9c0f351",
   "metadata": {},
   "source": [
    "#### Example - GENCODE gene annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "097a0780",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>biotype</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.5;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid  source     biotype  start    end score strand phase  \\\n",
       "0  chr1  HAVANA        gene  11869  14409     .      +     .   \n",
       "1  chr1  HAVANA  transcript  11869  14409     .      +     .   \n",
       "2  chr1  HAVANA        exon  11869  12227     .      +     .   \n",
       "3  chr1  HAVANA        exon  12613  12721     .      +     .   \n",
       "4  chr1  HAVANA        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...  \n",
       "1  ID=ENST00000456328.2;Parent=ENSG00000223972.5;...  \n",
       "2  ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3  ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4  ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filepath = resource_dir / \"general\" / \"gene_annotations.hg38.gencode_v40\" / \"gencode.v40.annotation.gff3.gz\"\n",
    "\n",
    "tmp1 = pd.read_csv(filepath, sep=\"\\t\", compression=\"gzip\", header=None, comment=\"#\", nrows=100, names=gff_columns)\n",
    "display(tmp1.head(5))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6467c007",
   "metadata": {},
   "source": [
    "Additionally, metadata files were generated to help processing these files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f35cffb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapping gene IDs to transcript ID(s), gene name, etc.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>official_name</th>\n",
       "      <th>official_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>HGNC:38034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>HGNC:50039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>HGNC:52482</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id            gene_id    gene_name  \\\n",
       "0  ENST00000456328.2  ENSG00000223972.5      DDX11L1   \n",
       "1  ENST00000450305.2  ENSG00000223972.5      DDX11L1   \n",
       "2  ENST00000488147.1  ENSG00000227232.5       WASH7P   \n",
       "3  ENST00000619216.1  ENSG00000278267.1    MIR6859-1   \n",
       "4  ENST00000473358.1  ENSG00000243485.5  MIR1302-2HG   \n",
       "\n",
       "                            gene_type protein_id official_name official_id  \n",
       "0  transcribed_unprocessed_pseudogene        NaN       DDX11L1  HGNC:37102  \n",
       "1  transcribed_unprocessed_pseudogene        NaN       DDX11L1  HGNC:37102  \n",
       "2              unprocessed_pseudogene        NaN        WASH7P  HGNC:38034  \n",
       "3                               miRNA        NaN     MIR6859-1  HGNC:50039  \n",
       "4                              lncRNA        NaN   MIR1302-2HG  HGNC:52482  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Mapping gene types to full descriptions.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genetype</th>\n",
       "      <th>Definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>IG_J_pseudogene</td>\n",
       "      <td>Inactivated immunoglobulin gene.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>miRNA_pseudogene</td>\n",
       "      <td>Non-coding RNA predicted to be pseudogene by t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>non_coding</td>\n",
       "      <td>Transcript which is known from the literature ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Contains an open reading frame (ORF).</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>retained_intron</td>\n",
       "      <td>Alternatively spliced transcript believed to c...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            genetype                                         Definition\n",
       "11   IG_J_pseudogene                   Inactivated immunoglobulin gene.\n",
       "34  miRNA_pseudogene  Non-coding RNA predicted to be pseudogene by t...\n",
       "42        non_coding  Transcript which is known from the literature ...\n",
       "39    protein_coding              Contains an open reading frame (ORF).\n",
       "38   retained_intron  Alternatively spliced transcript believed to c..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Table mapping gene types to simplified groups.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genetype</th>\n",
       "      <th>simplified_genetype</th>\n",
       "      <th>group_genetype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ribozyme</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>sRNA</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IG_J_gene</td>\n",
       "      <td>IG_TR_gene</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>translated_unprocessed_pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>translated_processed_pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             genetype simplified_genetype  group_genetype\n",
       "23                           ribozyme               ncRNA           ncRNA\n",
       "24                               sRNA               ncRNA           ncRNA\n",
       "2                           IG_J_gene          IG_TR_gene  protein_coding\n",
       "56  translated_unprocessed_pseudogene          pseudogene      pseudogene\n",
       "55    translated_processed_pseudogene          pseudogene      pseudogene"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filepath = resource_dir / \"general\" / \"gene_annotations.hg38.gencode_v40\" / \"gene_metadata.tsv.gz\"\n",
    "\n",
    "tmp1 = pd.read_csv(filepath, sep=\"\\t\", compression=\"gzip\", header=0)\n",
    "\n",
    "print(\"Mapping gene IDs to transcript ID(s), gene name, etc.\")\n",
    "display(tmp1.head(5))\n",
    "\n",
    "###\n",
    "print(\"\\n\")\n",
    "\n",
    "filepath = resource_dir / \"general\" / \"gene_annotations.hg38.gencode_v40\" / \"genetype_metadata.tsv\"\n",
    "\n",
    "print(\"Mapping gene types to full descriptions.\")\n",
    "tmp1 = pd.read_csv(filepath, sep=\"\\t\", header=0)\n",
    "display(tmp1.sample(5).sort_values(by=\"genetype\"))\n",
    "\n",
    "\n",
    "###\n",
    "print(\"\\n\")\n",
    "\n",
    "filepath = resource_dir / \"general\" / \"gene_annotations.hg38.gencode_v40\" / \"simplified_grouped_genetype.tsv\"\n",
    "\n",
    "print(\"Table mapping gene types to simplified groups.\")\n",
    "tmp1 = pd.read_csv(filepath, sep=\"\\t\", header=0)\n",
    "display(tmp1.sample(5).sort_values(by=\"group_genetype\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "972f1586",
   "metadata": {},
   "source": [
    "Finally, a set of files with non-overlapping intervals were generated (one for each strand). \n",
    "\n",
    "These intervals provide for each segment its most important \"biotype\", i.e. the type of gene-related annotation representing the most important biological function at this genomic interval.\n",
    "\n",
    "e.g. if two genes overlap at a given interval, where one gene has a CDS (coding sequence) annotated while the other gene has a UTR (untranslated region) annotated, the CDS will be considered as the most important biotype at this interval.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7090f1be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450739</td>\n",
       "      <td>451675</td>\n",
       "      <td>stop_codon;OR4F29;protein_coding;ENSG000002847...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>451675</td>\n",
       "      <td>451678</td>\n",
       "      <td>start_codon;OR4F29;protein_coding;ENSG00000284...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685715</td>\n",
       "      <td>686651</td>\n",
       "      <td>stop_codon;OR4F16;protein_coding;ENSG000002846...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>686651</td>\n",
       "      <td>686654</td>\n",
       "      <td>start_codon;OR4F16;protein_coding;ENSG00000284...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944202</td>\n",
       "      <td>944203</td>\n",
       "      <td>three_prime_UTR;NOC2L;protein_coding;ENSG00000...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>944204</td>\n",
       "      <td>three_prime_UTR;NOC2L;protein_coding;ENSG00000...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944204</td>\n",
       "      <td>944693</td>\n",
       "      <td>three_prime_UTR;NOC2L;protein_coding;ENSG00000...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944693</td>\n",
       "      <td>944696</td>\n",
       "      <td>stop_codon;NOC2L;protein_coding;ENSG0000018897...</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944696</td>\n",
       "      <td>944800</td>\n",
       "      <td>CDS;NOC2L;protein_coding;ENSG00000188976.11</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944800</td>\n",
       "      <td>945041</td>\n",
       "      <td>intron;NOC2L;protein_coding;ENSG00000188976.11</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom   start     end                                               name  \\\n",
       "0  chr1  450739  451675  stop_codon;OR4F29;protein_coding;ENSG000002847...   \n",
       "1  chr1  451675  451678  start_codon;OR4F29;protein_coding;ENSG00000284...   \n",
       "2  chr1  685715  686651  stop_codon;OR4F16;protein_coding;ENSG000002846...   \n",
       "3  chr1  686651  686654  start_codon;OR4F16;protein_coding;ENSG00000284...   \n",
       "4  chr1  944202  944203  three_prime_UTR;NOC2L;protein_coding;ENSG00000...   \n",
       "5  chr1  944203  944204  three_prime_UTR;NOC2L;protein_coding;ENSG00000...   \n",
       "6  chr1  944204  944693  three_prime_UTR;NOC2L;protein_coding;ENSG00000...   \n",
       "7  chr1  944693  944696  stop_codon;NOC2L;protein_coding;ENSG0000018897...   \n",
       "8  chr1  944696  944800        CDS;NOC2L;protein_coding;ENSG00000188976.11   \n",
       "9  chr1  944800  945041     intron;NOC2L;protein_coding;ENSG00000188976.11   \n",
       "\n",
       "  score strand  \n",
       "0     .      -  \n",
       "1     .      -  \n",
       "2     .      -  \n",
       "3     .      -  \n",
       "4     .      -  \n",
       "5     .      -  \n",
       "6     .      -  \n",
       "7     .      -  \n",
       "8     .      -  \n",
       "9     .      -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65418</td>\n",
       "      <td>65433</td>\n",
       "      <td>five_prime_UTR;OR4F5;protein_coding;ENSG000001...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65433</td>\n",
       "      <td>65519</td>\n",
       "      <td>intron;OR4F5;protein_coding;ENSG00000186092.7</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65519</td>\n",
       "      <td>65564</td>\n",
       "      <td>five_prime_UTR;OR4F5;protein_coding;ENSG000001...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65564</td>\n",
       "      <td>65567</td>\n",
       "      <td>start_codon;OR4F5;protein_coding;ENSG000001860...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65567</td>\n",
       "      <td>65573</td>\n",
       "      <td>CDS;OR4F5;protein_coding;ENSG00000186092.7</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65573</td>\n",
       "      <td>69036</td>\n",
       "      <td>intron;OR4F5;protein_coding;ENSG00000186092.7</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69036</td>\n",
       "      <td>70005</td>\n",
       "      <td>CDS;OR4F5;protein_coding;ENSG00000186092.7</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70005</td>\n",
       "      <td>70008</td>\n",
       "      <td>stop_codon;OR4F5;protein_coding;ENSG00000186092.7</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>70008</td>\n",
       "      <td>71585</td>\n",
       "      <td>three_prime_UTR;OR4F5;protein_coding;ENSG00000...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923922</td>\n",
       "      <td>924431</td>\n",
       "      <td>five_prime_UTR;SAMD11;protein_coding;ENSG00000...</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom   start     end                                               name  \\\n",
       "0  chr1   65418   65433  five_prime_UTR;OR4F5;protein_coding;ENSG000001...   \n",
       "1  chr1   65433   65519      intron;OR4F5;protein_coding;ENSG00000186092.7   \n",
       "2  chr1   65519   65564  five_prime_UTR;OR4F5;protein_coding;ENSG000001...   \n",
       "3  chr1   65564   65567  start_codon;OR4F5;protein_coding;ENSG000001860...   \n",
       "4  chr1   65567   65573         CDS;OR4F5;protein_coding;ENSG00000186092.7   \n",
       "5  chr1   65573   69036      intron;OR4F5;protein_coding;ENSG00000186092.7   \n",
       "6  chr1   69036   70005         CDS;OR4F5;protein_coding;ENSG00000186092.7   \n",
       "7  chr1   70005   70008  stop_codon;OR4F5;protein_coding;ENSG00000186092.7   \n",
       "8  chr1   70008   71585  three_prime_UTR;OR4F5;protein_coding;ENSG00000...   \n",
       "9  chr1  923922  924431  five_prime_UTR;SAMD11;protein_coding;ENSG00000...   \n",
       "\n",
       "  score strand  \n",
       "0     .      +  \n",
       "1     .      +  \n",
       "2     .      +  \n",
       "3     .      +  \n",
       "4     .      +  \n",
       "5     .      +  \n",
       "6     .      +  \n",
       "7     .      +  \n",
       "8     .      +  \n",
       "9     .      +  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# NOTE: we provide the comprehensive set :\n",
    "\n",
    "\n",
    "selected_annotation_set: Literal[\"complete\", \"proteincoding_only\"] = \"proteincoding_only\"\n",
    "\n",
    "filepath_minus = (\n",
    "    resource_dir\n",
    "    / \"general\"\n",
    "    / \"gene_annotations.hg38.gencode_v40\"\n",
    "    / f\"{selected_annotation_set}.non-overlap.annotated.minus.bed.gz\"\n",
    ")\n",
    "filepath_plus = (\n",
    "    resource_dir\n",
    "    / \"general\"\n",
    "    / \"gene_annotations.hg38.gencode_v40\"\n",
    "    / f\"{selected_annotation_set}.non-overlap.annotated.plus.bed.gz\"\n",
    ")\n",
    "\n",
    "tmp1 = pd.read_csv(filepath_minus, sep=\"\\t\", compression=\"gzip\", header=None, comment=\"#\", names=bed6_cols, nrows=100)\n",
    "tmp2 = pd.read_csv(filepath_plus, sep=\"\\t\", compression=\"gzip\", header=None, comment=\"#\", names=bed6_cols, nrows=100)\n",
    "\n",
    "display(tmp1.head(10))\n",
    "display(tmp2.head(10))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da4e4183",
   "metadata": {},
   "source": [
    "### Genome sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f72c0da0",
   "metadata": {},
   "source": [
    "Nucleotide (and protein) sequences can be stored in text files following the FASTA format.\n",
    "\n",
    "Such files can be compressed with `bgzip` and indexed with `samtools faidx` to enable fast access to specific sequences.\n",
    "\n",
    "We can then use dedicated libraries to query the file, e.g. to retrieve the sequence of a specific genomic interval.\n",
    "\n",
    "Here with the `pyfaidx` library, we can retrieve the sequence of a specific genomic interval."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ba65ba0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "fa = pyfaidx.Fasta(resource_dir / \"general\" / \"genome\" / \"hg38.fa.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "500f08e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Protocol\n",
    "\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class GenomicInterval:\n",
    "    chrom: str\n",
    "    start: int\n",
    "    end: int\n",
    "    name: str\n",
    "    score: int | float | str\n",
    "    strand: Literal[\"+\", \"-\"]\n",
    "\n",
    "    def __post_init__(self):\n",
    "        # Enforce [0, end) interval.\n",
    "        if self.start < 0:\n",
    "            raise ValueError(f\"Start position {self.start} must be non-negative.\")\n",
    "        if self.start >= self.end:\n",
    "            raise ValueError(f\"Start position {self.start} must be less than end position {self.end}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e5ff851",
   "metadata": {},
   "source": [
    "#### Example - Plus (or \"forward\") strand sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b8fa3dbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCCAGATCTCTTCAG\n",
      "Verify that length of query result is correct:\n",
      "15 == 15? True\n",
      "\n",
      "\n",
      "Verify that the query result is the same for both syntaxes:\n",
      "(query_result == query_result_alt)=True\n",
      "CCCAGATCTCTTCAG\n",
      "CCCAGATCTCTTCAG\n"
     ]
    }
   ],
   "source": [
    "# First example is on the FORWARD, PLUS strand.\n",
    "example_interval = GenomicInterval(\n",
    "    chrom=\"chr1\", start=65418, end=65433, name=\"five_prime_UTR;OR4F5\", score=\".\", strand=\"+\"\n",
    ")\n",
    "\n",
    "# CAUTION : query is 1-based, inclusive at end.\n",
    "query_result = str(\n",
    "    fa.get_seq(\n",
    "        name=example_interval.chrom,\n",
    "        start=example_interval.start + 1,\n",
    "        end=example_interval.end,\n",
    "        rc=example_interval.strand == \"-\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# To simplify : one can use a more classic slicing syntax, with 0-based indexing.\n",
    "query_result_alt = fa[\"chr1\"][example_interval.start : example_interval.end]\n",
    "print(query_result_alt)\n",
    "\n",
    "print(\"Verify that length of query result is correct:\")\n",
    "print(\n",
    "    f\"{len(query_result):_} == {example_interval.end - example_interval.start:_}? \"\n",
    "    f\"{len(query_result) == example_interval.end - example_interval.start}\"\n",
    ")\n",
    "\n",
    "print(\"\\n\")\n",
    "print(\"Verify that the query result is the same for both syntaxes:\")\n",
    "print(f\"{(query_result == query_result_alt)=}\")\n",
    "print(query_result)\n",
    "print(query_result_alt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19627d32",
   "metadata": {},
   "source": [
    "#### Example - Minus (or \"reverse\") strand sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0481c8e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GenomicInterval(chrom='chr1', start=944204, end=944224, name='(last 20 nt) three_prime_UTR;NOC2L', score='.', strand='-')\n",
      "\n",
      "\n",
      "Query result:\n",
      ">chr1:944224-944205 (complement)\n",
      "GGAGGCAGGGCCATTGTGTT\n",
      "\n",
      "\n",
      "Checking the alternative syntax:\n",
      "(query_result_minus_alt == query_result_minus)=True\n",
      "GGAGGCAGGGCCATTGTGTT\n",
      "(len(query_result_minus)== (example_interval.end - example_interval.start))=True\n",
      "\n",
      "\n",
      "Displaying the fact that the sequence is a reverse-complement of the forward strand:\n",
      "5'  A-A-C-A-C-A-A-T-G-G...  3'\n",
      "    | | | | | | | | | | \n",
      "3'  T-T-G-T-G-T-T-A-C-C...  5'\n"
     ]
    }
   ],
   "source": [
    "# Second example is on the REVERSE, MINUS strand.\n",
    "example_interval = GenomicInterval(\n",
    "    chrom=\"chr1\", start=944204, end=944224, name=\"(last 20 nt) three_prime_UTR;NOC2L\", score=\".\", strand=\"-\"\n",
    ")\n",
    "\n",
    "print(example_interval)\n",
    "\n",
    "# CAUTION : query is 1-based, inclusived at end.\n",
    "query_result_minus = fa.get_seq(\n",
    "    name=example_interval.chrom,\n",
    "    start=example_interval.start + 1,\n",
    "    end=example_interval.end,\n",
    "    rc=example_interval.strand == \"-\",\n",
    ")\n",
    "\n",
    "print(\"\\n\")\n",
    "print(\"Query result:\")\n",
    "print(query_result_minus.__repr__())\n",
    "\n",
    "# We can test the alternative syntax. Here we apply the two methods to reverse and complement.\n",
    "query_result_minus_alt = (fa[example_interval.chrom][example_interval.start : example_interval.end]).reverse.complement\n",
    "\n",
    "###\n",
    "\n",
    "print(\"\\n\")\n",
    "print(\"Checking the alternative syntax:\")\n",
    "print(f\"{(query_result_minus_alt == query_result_minus)=}\")\n",
    "\n",
    "print(query_result_minus_alt)\n",
    "print(f\"{(len(query_result_minus)== (example_interval.end - example_interval.start))=}\")\n",
    "\n",
    "###\n",
    "\n",
    "# We can make clear that the sequence is indeed a reverse-complement of the forward strand.\n",
    "\n",
    "map_complement = {\"A\": \"T\", \"T\": \"A\", \"C\": \"G\", \"G\": \"C\", \"N\": \"N\"}\n",
    "query_result_plus = fa[example_interval.chrom][example_interval.start : example_interval.end]\n",
    "\n",
    "print(\"\\n\")\n",
    "print(\"Displaying the fact that the sequence is a reverse-complement of the forward strand:\")\n",
    "print(\"5'  \" + \"-\".join(str(query_result_plus)[:10]) + \"...\" + \"  3'\")\n",
    "print(\n",
    "    \"    \"\n",
    "    + \"\".join(\n",
    "        [\"| \" for a, b in zip(str(query_result_plus), str(query_result_minus)[::-1]) if map_complement[a] == b][:10]\n",
    "    )\n",
    ")\n",
    "print(\"3'  \" + \"-\".join(str(query_result_minus)[-10:][::-1]) + \"...\" + \"  5'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7388cff2",
   "metadata": {},
   "source": [
    "## Integrating everything: intervals and sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea2d4072",
   "metadata": {},
   "source": [
    "Another library that can be used to deal with biology-related data is BioPython.\n",
    "\n",
    "Notably here, we can use it to deal with FASTA sequences.\n",
    "\n",
    "Here, we will :\n",
    "\n",
    "1. extract all the regions for a given transcript from the MANE annotations.\n",
    "2. focus on the CDS (coding sequence) intervals.\n",
    "3. query the FASTA file to retrieve the sequence of these intervals.\n",
    "4. concatenate them and translate the resulting sequence into a protein sequence.\n",
    "5. export to a mock output file the CDS sequence intervals.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb9f9e4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65564</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS;ENST00000641515.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69036</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS;ENST00000641515.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  start    end score strand                   name\n",
       "4    chr1  65564  65573     .      +  CDS;ENST00000641515.2\n",
       "7    chr1  69036  70008     .      +  CDS;ENST00000641515.2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Concatenated sequence length: 981\n",
      "len(concatenated_seq.translate())=327\n",
      "\n",
      "\n",
      "Demonstration: reading back the exported SeqRecords.\n",
      "CDS;ENST00000641515.2 CDS;ENST00000641515.2 chrom=chr1, start=65565, end=65573, strand=+, score=. 9\n",
      "CDS;ENST00000641515.2 CDS;ENST00000641515.2 chrom=chr1, start=69037, end=70008, strand=+, score=. 972\n"
     ]
    }
   ],
   "source": [
    "### STEP 1\n",
    "\n",
    "gff_filepath = resource_dir / \"general\" / \"gene_annotations.hg38.MANE_v1.4\" / \"MANE.GRCh38.v1.4.ensembl_genomic.gff.gz\"\n",
    "bed_filepath = (\n",
    "    resource_dir / \"general\" / \"gene_annotations.hg38.MANE_v1.4\" / \"MANE.GRCh38.v1.4.ensembl_genomic.transcripts.bed.gz\"\n",
    ")\n",
    "\n",
    "bed_mane = pd.read_csv(\n",
    "    bed_filepath,\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\",\n",
    "    header=None,\n",
    "    comment=\"#\",\n",
    "    names=bed6_cols,\n",
    "    nrows=1,\n",
    ")\n",
    "\n",
    "\n",
    "# NOTE: here we build a BedTool object directly from the filepath.\n",
    "# BedTool is able to detect it is a GFF format and will parse it accordingly.\n",
    "bt_gff_mane = pbt.BedTool(gff_filepath)\n",
    "bt_bed_mane = pbt.BedTool.from_dataframe(bed_mane.iloc[[0], :])\n",
    "\n",
    "gff_extract = bt_gff_mane.intersect(bt_bed_mane, s=True).to_dataframe().sort_values(by=[\"start\", \"end\"])\n",
    "\n",
    "### STEP 2\n",
    "\n",
    "# We now extract the intervals with CDS biotypes.\n",
    "cds_gff = gff_extract[gff_extract[\"feature\"] == \"CDS\"].copy()\n",
    "\n",
    "# Reminder: GFF are 1-based, inclusive intervals. So to convert into 0-based,\n",
    "# you simply subtract 1 from the start position.\n",
    "\n",
    "cds_bed = cds_gff.loc[:, [\"seqname\", \"start\", \"end\", \"score\", \"strand\"]].copy()\n",
    "cds_bed[\"start\"] = cds_bed[\"start\"] - 1  # Convert to 0-based start position.\n",
    "cds_bed[\"name\"] = cds_gff[\"feature\"] + \";\" + bed_mane[\"name\"].values[0]  # Add transcript name to CDS name.\n",
    "\n",
    "display(cds_bed.head(3))\n",
    "\n",
    "### STEP 3 - get sequences\n",
    "\n",
    "seqrecords: list[SeqRecord] = []\n",
    "\n",
    "for _, row in cds_bed.iterrows():\n",
    "    seq = fa.get_seq(name=row[\"seqname\"], start=row[\"start\"] + 1, end=row[\"end\"], rc=row[\"strand\"] == \"-\")\n",
    "\n",
    "    # Make a Bio.SeqRecord object for each sequence.\n",
    "    seq = Seq(str(seq))\n",
    "    # Convert to RNA.\n",
    "    seq = seq.transcribe()\n",
    "\n",
    "    if len(seq) != (row[\"end\"] - row[\"start\"]):\n",
    "        raise ValueError(\n",
    "            f\"Length of sequence {len(seq)} does not match expected length {row['end'] - row['start']} for \"\n",
    "            f\"chrom={row['seqname']}, start={row['start'] + 1}, end={row['end']}, strand={row['strand']}.\"\n",
    "        )\n",
    "\n",
    "    seqrec = SeqRecord(\n",
    "        Seq(str(seq)),\n",
    "        id=row[\"name\"],\n",
    "        description=f\"chrom={row['seqname']}, start={row['start'] + 1}, end={row['end']}, strand={row['strand']}, score={score}\",\n",
    "    )\n",
    "\n",
    "    seqrecords.append(seqrec)\n",
    "\n",
    "\n",
    "### STEP 4 - Example of translation of the sequence with BioPython.\n",
    "\n",
    "# We concatenate the seq records into a single sequence.\n",
    "concatenated_seq = Seq(\"\".join([str(seqrec.seq) for seqrec in seqrecords]))\n",
    "print(\"Concatenated sequence length:\", len(concatenated_seq))\n",
    "print(f\"{len(concatenated_seq.translate())=}\")\n",
    "\n",
    "\n",
    "### STEP 5 - Export to fake FASTA file, and read it back.\n",
    "\n",
    "fasta_output = StringIO()\n",
    "SeqIO.write(seqrecords, fasta_output, \"fasta\")\n",
    "fasta_output.seek(0)  # Reset the StringIO object to the beginning.\n",
    "\n",
    "print(\"\\n\")\n",
    "print(\"Demonstration: reading back the exported SeqRecords.\")\n",
    "reader = SeqIO.parse(fasta_output, \"fasta\")  # Read the first record from the StringIO object.\n",
    "for seqrec in reader:\n",
    "    print(seqrec.id, seqrec.description, len(seqrec.seq))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "parnet-demo-gpu",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
